raster/src/main/scala/geotrellis/raster/ForEachCell.scala

```diff
@@ -0,0 +1,134 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package geotrellis.raster
+
+import geotrellis.raster._
+import geotrellis.raster.rasterize.{Rasterizer, Callback}
+import geotrellis.vector.Geometry
+
+
+/**
+ * For Geometry Objects
+  */
+trait GeometryForEachCellMethods[T <: Geometry] extends MethodExtensions[T] {
+  import geotrellis.raster.rasterize.{Transformer,Rasterizer}
+
+  def foreachCell(re : RasterExtent, includeExterior : Boolean = false)(fn : Int => Int) : Tile =
+    re.foreachCell(self, includeExterior)(fn)
+
+  def foreachCellDouble(re : RasterExtent, includeExterior : Boolean = false)(fn : Double => Double) : Tile =
+    re.foreachCellDouble(self, includeExterior)(fn)
+
+  def rasterize(re : RasterExtent)(fn : Transformer[Int]) = Rasterizer.rasterize(self, re)(fn)
+
+  def foreachCellCovering(re : RasterExtent)(fn : Int => Int) : Tile = foreachCell(re, true)(fn)
+  def foreachCellContanedBy(re : RasterExtent)(fn : Int => Int) : Tile = foreachCell(re, false)(fn)
+  def foreachCellCoveringDouble(re : RasterExtent)(fn : Double => Double) : Tile = foreachCellDouble(re, true)(fn)
+  def foreachCellContanedByDouble(re : RasterExtent)(fn : Double => Double) : Tile = foreachCellDouble(re, false)(fn)
+}
+
+/**
+ * For Raster Extents
+ */
+trait RasterExtentForEachCellMethods[T <: RasterExtent] extends MethodExtensions[T] {
+  def foreachCell(geom : Geometry, includeExterior : Boolean = false)(fn : Int => Int) : Tile = {
+    val nd = NODATA
+    val mutableTile = IntArrayTile.fill(nd, self.cols, self.rows)
+
+    Rasterizer.foreachCellByGeometry(geom, self, includeExterior)( new Callback {
+      def apply(col : Int, row : Int): Unit = mutableTile.set(col, row, fn(nd))
+    })
+    mutableTile
+  }
+
+  def foreachCellDouble(geom : Geometry, includeExterior : Boolean = false)(fn : Double => Double) : Tile = {
+    val nd = doubleNODATA
+    val mutableTile = DoubleArrayTile.fill(nd, self.cols, self.rows)
+
+    Rasterizer.foreachCellByGeometry(geom, self, includeExterior)( new Callback {
+      def apply(col : Int, row : Int): Unit = mutableTile.setDouble(col, row, fn(nd))
+    })
+    mutableTile
+  }
+
+  def foreachCellCovering(geom : Geometry)(fn : Int => Int) : Tile = foreachCell(geom, true)(fn)
+  def foreachCellContainedBy(geom : Geometry)(fn : Int => Int) : Tile = foreachCell(geom, false)(fn)
+  def foreachCellCoveringDouble(geom : Geometry)(fn : Double => Double) : Tile = foreachCellDouble(geom, true)(fn)
+  def foreachCellContainedByDouble(geom : Geometry)(fn : Double => Double) : Tile = foreachCellDouble(geom, false)(fn)
+}
+
+/**
+ * For Tiles
+ */
+trait TileForEachCellMethods[T <: Tile] extends MethodExtensions[T] {
+  def foreachCell(geom : Geometry, re : RasterExtent, includeExterior : Boolean = false)(fn : Int => Int) : Tile = {
```

Comment:
Better to write this against `Raster`, so that the extent is already there. The tile methods can then take an extent (doesn't require a RasterExtent, since it already has it's own cols and rows), and then delegate to a Raster method by `Raster(tile, extent).foreach(...)`